sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala

```diff
@@ -355,11 +355,22 @@ class StreamingQueryManager private[sql] (sparkSession: SparkSession) extends Lo
       // Make sure no other query with same id is active across all sessions
       val activeOption =
         Option(sparkSession.sharedState.activeStreamingQueries.putIfAbsent(query.id, this))
-      if (activeOption.isDefined || activeQueries.values.exists(_.id == query.id)) {
+
+      val streamAlreadyActive =
+        activeOption.isDefined || activeQueries.values.exists(_.id == query.id)
+      val turnOffOldStream =
+        sparkSession.sessionState.conf.getConf(SQLConf.STOP_RUNNING_DUPLICATE_STREAM)
+      if (streamAlreadyActive && turnOffOldStream) {
+        val queryManager = activeOption.getOrElse(this)
+        logInfo(s"Stopping existing streaming query [id=${query.id}], as a new run is being " +
+          "started.")
+        queryManager.get(query.id).stop()
```

Comment:
If the existing stream is a "zombie", can it happen that it does not respond to stop() and then this will block forever?

Reply:
Great question. I can add some safeguards against this, but in most cases we mean that the stream is a "zombie", because we lost all references to it, not because it is uninterruptable.

Reply:
I think this should be fine. If `stop` returns, the query should already be stopped, because `stop` waits until the streaming thread dies.

Reply:
Ah, this has a deadlock. We are waiting for stopping query inside a lock that query needs it to remove it from the active queries.